jwst/badpix_selfcal/badpix_selfcal_step.py

```diff
@@ -0,0 +1,210 @@
+        association file, including science, background, and selfcal exposures,
+        are included in the MIN frame from which outliers are detected.
+        If selfcal_list and/or bkg_list are specified manually, they are appended to any
+        selfcal or background exposures found in the input association file.
+        If selfcal_list and bkg_list are both set to None and input is
+        a single science exposure, the step will be skipped with a warning unless
+        the force_single parameter is set True.
+        In that case, the input exposure will be used as the sole background exposure,
+        i.e., true self-calibration.
+        """
+        input_sci, selfcal_list, bkg_list = _parse_inputs(input, selfcal_list, bkg_list)
+
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list + bkg_list) == 0) and (not self.force_single):
+            self.log.warning("No selfcal or background exposures provided for self-calibration. "
+                        "Skipping step. If you want to force self-calibration with the science "
+                        "exposure alone (generally not recommended), set force_single=True.")
+            input_sci.meta.cal_step.badpix_selfcal = 'SKIPPED'
+            return input_sci, bkg_list
+
+        # get the dispersion axis
+        try:
+            dispaxis = input_sci.meta.wcsinfo.dispersion_direction
+        except AttributeError:
+            self.log.warning("Dispersion axis not found in input science image metadata. "
+                        "Kernel for self-calibration will be two-dimensional.")
+            dispaxis = None
+
+        # collapse all selfcal exposures into a single background model
+        # note that selfcal_list includes the science exposure. This is expected.
+        # all exposures are combined into a single background model using a MIN operation.
+        selfcal_list = [input_sci] + selfcal_list
+        selfcal_3d = []
+        for i, selfcal_model in enumerate(selfcal_list):
+            selfcal_3d.append(selfcal_model.data)
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=RuntimeWarning, message="All-NaN")
+            minimg = np.nanmin(np.asarray(selfcal_3d), axis=0)
+        bad_indices = badpix_selfcal.badpix_selfcal(minimg, self.flagfrac_lower, self.flagfrac_upper, self.kernel_size, dispaxis)
+
+        # apply the flags to the science data
+        input_sci = badpix_selfcal.apply_flags(input_sci, bad_indices)
+
+        # apply the flags to the background data to be passed to background sub step
+        if len(bkg_list) > 0:
+            for i, background_model in enumerate(bkg_list):
+                bkg_list[i] = badpix_selfcal.apply_flags(dm.open(background_model), bad_indices)
+
+        if self.save_flagged_bkg:
+            self.save_bkg(bkg_list)
+
+        input_sci.meta.cal_step.badpix_selfcal = 'COMPLETE'
```

Comment:
Has an update been made to stdatamodels core schema to define the new `cal_step.badpix_selfcal` attribute? At some point in the future we'll also need to file a JWSTKD ticket to add the new keyword to the official dictionary. At this point it's not critical, because the step is skipped by default and hence the keyword won't be used in products coming out of operations.

Reply:
You should go ahead and submit a PR against stdatamodels to at least get it added into our schema. I'll give myself the action to file a JWSTKD ticket to have it added to the jwst keyword dictionary.

Reply:
Sounds good.  That [PR is here](https://github.com/spacetelescope/stdatamodels/pull/305/files).